src/snmalloc/global/memcpy.h

```diff
@@ -459,4 +469,42 @@ namespace snmalloc
     Arch::copy(dst, src, len);
     return orig_dst;
   }
+
+  template<
+    bool Checked,
+    bool ReadsChecked = CheckReads,
+    typename Arch = DefaultArch>
+  SNMALLOC_FAST_PATH_INLINE void*
+  memmove(void* dst, const void* src, size_t len)
+  {
+    auto orig_dst = dst;
+    // we don't need to do external
+    // pointer checks if we hit it.  It's also the fastest case, to encourage
+    // the compiler to favour the other cases.
+    if (SNMALLOC_UNLIKELY(len == 0 || dst == src))
+    {
+      return dst;
+    }
+
+    // Check the bounds of the arguments.
+    if (SNMALLOC_UNLIKELY(!check_bounds<(Checked && ReadsChecked)>(src, len)))
+      return report_fatal_bounds_error(
+        src, len, "memmove with source out of bounds of heap allocation");
+    if (SNMALLOC_UNLIKELY(!check_bounds<Checked>(dst, len)))
+      return report_fatal_bounds_error(
+        dst, len, "memmove with destination out of bounds of heap allocation");
+
+    if (dst > src)
```

Comment:
Would 
```suggestion
    if ((dst - src) < len)
```
be a better check?  Currently it doesn't check that there is actually an overlap?  @davidchisnall

Reply:
I guess my concern here is because you have dropped to a slow byte by byte copy.  Half of all non-overlapping `memmove`s will be slow.

Reply:
right .. I ll follow your suggestion.